### ISSUE TITLE
Question between Call(x,self) and Dummy_loss(y_true, y_pred) function

### ISSUE THREAD

**Issue by esconda** [experiencor/keras-yolo3#159]
Hi,
I just would like to ask, you use dummy loss in model compile function as loss function as : **train_model.compile(loss=dummy_loss, optimizer=optimizer)**

But According to dummy loss function , it does not calculate any loss.It directly returns : **tf.sqrt(tf.reduce_sum(y_pred))**

Your actual loss calculation is available in **Call(x,self)** function.But why dont you use call function in compile rather than dumm_loss function.

I tried it with pretrained yolo_v3 weights and it worked well with dummy loss function.
I checked all pages and looked for Call(self,x) function but unfortunately you didn't use this function in anywhere of your code.

But I really didn't understand how it worked without of calculating loss function which is available in **Call(self,x)** .

Could you please explain how dummy_loss worked or why you didn't call Call(self,x) in compile as : **train_model.compile(loss=Call optimizer=optimizer)**

**Comment by NanoLsk** [comment]
> Hi,
> I just would like to ask, you used dummy_loss function in model compile function as loss function: **train_model.compile(loss=dummy_loss, optimizer=optimizer)**
> 
> But According to dummy loss function , it does not calculate any loss.It directly returns : **tf.sqrt(tf.reduce_sum(y_pred))**
> 
> Your actual loss calculation is available in **Call(self,x)** function.But why didn't you use this Call(self,x) function in compile rather than dummy_loss function.
> 
> I tried it with pretrained yolo_v3 weights and it worked well with dummy loss function.
> I checked all pages and looked for Call(self,x) function but unfortunately you didn't use this function in anywhere of your code.
> 
> But I really didn't understand how it worked without of calculating loss function which is available in **Call(self,x)** .
> 
> Could you please explain how dummy_loss worked or why you didn't use Call(self,x) in compile as : **train_model.compile(loss=Call optimizer=optimizer)**
can you solve this problem？ and i get the loss nan all the time.

**Comment by xxzfeng** [comment]
Hi,
I just would like to ask, you used dummy_loss function in model compile function as loss function: train_model.compile(loss=dummy_loss, optimizer=optimizer)

But According to dummy loss function , it does not calculate any loss.It directly returns : tf.sqrt(tf.reduce_sum(y_pred))

Your actual loss calculation is available in Call(self,x) function.But why didn't you use this Call(self,x) function in compile rather than dummy_loss function.

I tried it with pretrained yolo_v3 weights and it worked well with dummy loss function.
I checked all pages and looked for Call(self,x) function but unfortunately you didn't use this function in anywhere of your code.

But I really didn't understand how it worked without of calculating loss function which is available in Call(self,x) .

Could you please explain how dummy_loss worked or why you didn't use Call(self,x) in compile as : train_model.compile(loss=Call optimizer=optimizer)
can you solve this problem？ and i get the loss nan all the time.

**Comment by kabrau** [comment]
I'll try to explain, but it is my understanding, in the training, the last layer is loss only, the detection occurs before.
If you see the layers, the Ground Truth is inputted almost at the end of the net, and the box extraction and class definition of the prediction occur before the last layer, that is, the comparison between predict and ground truth occur at the penultimate layer.
(See create_yolov3_model and YoloLayer)
I hope I have helped you.

**Comment by paapu88** [comment]
Hi,
Unfortunately this issue is left pretty un-commented in the code. It has to do the fact that yolo3 neural net has three outputs (one for each grid) and in the training there should be a single loss function based on these three outputs. What I have gathered, standard keras cannot handle this. 

The way around seems to be in this code:
During the training, the three grids are trained independently on each other.
Loss is the 'dummy_loss' (in the end of yolo.py) which drives the fitting to minimize the prediction, which in case of training is sum of losses of those three grids. So minimizing dummy loss is the same as minimising the prediction (the output of the neural net) in the training phase (that is the sum of [loss_yolo_1, loss_yolo_2, loss_yolo_3] ).

You can see in the code, that the input and output of the neural net differs in case of training and prediction (in the end of yolo.py)
  

>   train_model = Model([input_image, true_boxes, true_yolo_1, true_yolo_2, true_yolo_3], [loss_yolo_1, loss_yolo_2, loss_yolo_3])
>     infer_model = Model(input_image, [pred_yolo_1, pred_yolo_2, pred_yolo_3])
> 

Terveisin, Markus
